modules/core/keeper/msg_server.go

```diff
@@ -751,7 +751,84 @@ func (k Keeper) ChannelUpgradeInit(goCtx context.Context, msg *channeltypes.MsgC
 
 // ChannelUpgradeTry defines a rpc handler method for MsgChannelUpgradeTry.
 func (k Keeper) ChannelUpgradeTry(goCtx context.Context, msg *channeltypes.MsgChannelUpgradeTry) (*channeltypes.MsgChannelUpgradeTryResponse, error) {
-	return nil, nil
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	module, _, err := k.ChannelKeeper.LookupModuleByChannel(ctx, msg.PortId, msg.ChannelId)
+	if err != nil {
+		ctx.Logger().Error("channel upgrade try failed", "port-id", msg.PortId, "error", errorsmod.Wrap(err, "could not retrieve module from port-id"))
+		return nil, errorsmod.Wrap(err, "could not retrieve module from port-id")
+	}
+	cbs, ok := k.Router.GetRoute(module)
+	if !ok {
+		ctx.Logger().Error("channel upgrade try failed", "port-id", msg.PortId, "error", errorsmod.Wrapf(porttypes.ErrInvalidRoute, "route not found to module: %s", module))
+		return nil, errorsmod.Wrapf(porttypes.ErrInvalidRoute, "route not found to module: %s", module)
+	}
+
+	channel, found := k.ChannelKeeper.GetChannel(ctx, msg.PortId, msg.ChannelId)
+	if !found {
+		return nil, errorsmod.Wrapf(channeltypes.ErrChannelNotFound, "port ID (%s) channel ID (%s)", msg.PortId, msg.ChannelId)
+	}
+
+	proposedUpgrade, err := k.ChannelKeeper.ChanUpgradeTry(
+		ctx,
+		msg.PortId,
+		msg.ChannelId,
+		msg.ProposedUpgradeFields,
+		msg.ProposedUpgradeTimeout,
+		msg.CounterpartyProposedUpgrade,
+		msg.CounterpartyUpgradeSequence,
+		msg.ProofChannel,
+		msg.ProofUpgrade,
+		msg.ProofHeight,
+	)
+	if err != nil {
+		ctx.Logger().Error("channel upgrade try failed", "error", errorsmod.Wrap(err, "channel handshake upgrade try failed"))
+		return &channeltypes.MsgChannelUpgradeTryResponse{
+			ChannelId:          msg.ChannelId,
+			Version:            proposedUpgrade.Fields.Version,
+			UpgradeSequence:    channel.UpgradeSequence,
+			ResponseResultType: channeltypes.FAILED,
+		}, err
+	}
+
+	proposedUpgradeVersion, err := cbs.OnChanUpgradeTry(
+		ctx,
+		msg.CounterpartyProposedUpgrade.Fields.Ordering,
+		msg.CounterpartyProposedUpgrade.Fields.ConnectionHops,
+		msg.PortId,
+		msg.ChannelId,
+		channel.UpgradeSequence,
+		channel.Counterparty,
+		channel.Version,
+		msg.CounterpartyProposedUpgrade.Fields.Version,
+	)
+	if err != nil {
+		ctx.Logger().Error("channel upgrade try callback failed", "port-id", msg.PortId, "channel-id", msg.ChannelId, "error", err.Error())
+		// TODO: commit error receipt to state and abort channel upgrade
+		return &channeltypes.MsgChannelUpgradeTryResponse{
+			ChannelId:          msg.ChannelId,
+			Version:            proposedUpgradeVersion,
+			UpgradeSequence:    channel.UpgradeSequence,
+			ResponseResultType: channeltypes.FAILED,
+		}, err
+	}
+
+	// set version to return value of callback
+	proposedUpgrade.Fields.Version = proposedUpgradeVersion
+
+	k.ChannelKeeper.WriteUpgradeTryChannel(
+		ctx,
+		msg.PortId,
+		msg.ChannelId,
+		channel,
```

Comment:
this is unsafe atm since the channel is obtained before the call to ChanUpgradeTry which will modify the channel upgrade sequence. I think for safety we should have each Write function get a fresh channel even if it is extra reads to disk